### PR TITLE
Add functions to set identifiers on AST node and edge patterns

### PR DESCRIPTION
**Pull request by jeffreylovitz**
Hi @cleishm,

This PR extends the libcypher-parser API such that identifiers can be added to nodes and relations after the AST has been created. We have found this to be a useful feature for standardizing our AST accesses.

I'm not aware of another place in this repository that modifies `children` arrays as we do here, which is necessary to properly free the added identifiers. Please let me know if anything should be changed to better match the project's style!

### CODE REVIEW

**Comment by cleishm** [comment]
Interesting! I'll take a look.

Generally, you may notice that the API creates an immutable AST and there are no exposed `set` methods to modify it after that. This would be the first such method that allows the modification of a node within an existing AST.

So I'd want to think a little on that. What's the use case that is driving that need?

Cheers,
Chris

File: lib/src/ast_node_pattern.c [hunk 1]
```diff
@@ -86,6 +86,29 @@ const cypher_astnode_t *cypher_ast_node_pattern_get_identifier(
...
+    node->identifier = identifier;
+    unsigned int n = astnode->nchildren;
+    if(n == 0)
+    {
+        astnode->children = malloc(sizeof(cypher_astnode_t *));
+    }
+    else
+    {
+        astnode->children =
+            realloc(astnode->children, (n + 1) * sizeof(cypher_astnode_t *));
```
Comment (cleishm) [inline lib/src/ast_node_pattern.c]: It's a bit pedantic, but I like to ensure there's always a check for a failed `malloc`/`realloc` (check for `NULL`) and then return an appropriate error indicator and set `errno`.

**Comment by jeffreylovitz** [comment]
Thanks for the quick reply!

For RedisGraph, this functionality serves a few useful purposes. The clearest example is probably the usage of inlined filtering conditions, as in:
`MATCH (:Person{name: 'Chris'})-[]->(r:Repository) WHERE r.stars > 25 RETURN r`

Because we perform traversals using matrix multiplication rather than following edges on adjacency lists, this traversal produces a matrix with nonzero values at every position that a `Person` has an outgoing relation with a `Repository`. Afterwards, we iterate over this matrix to see whether the 2 conditions (`name` and `stars`) are satisfied.

By adding identifiers to the AST, we can build separate Traversal and Filter operations that clearly work on the same graph entities:
```
1) Results
2)     Project
3)         Filter
4)             Conditional Traverse | (anon_0:Person)->(r:Repository)
```
Where `Filter` contains the constraints:
```
anon_0.name = Chris
r.stars > 25
```

Basically, enriching the AST allows all of our operations to be constructed with the set of identifiers that they operate upon, creating a common language for structuring the execution tree.

File: lib/src/ast_node_pattern.c (hunk 1 above)
Comment (jeffreylovitz) [reply lib/src/ast_node_pattern.c]: No problem, I'll add that tomorrow!

**Comment by cleishm** [comment]
Got it. My next question is whether that is best done by augmenting the tree using an alternative data structure, or by copying, rather than mutation.

If we do go the mutation route, the immediate question is whether a consistent API would mean needing to add mutators for a lot of other AST attributes?

**Comment by jeffreylovitz** [comment]
I'm totally open to discussing the various approaches available to us, but am uncertain of what alternative data structures you're considering!

Speaking from our experience on RedisGraph, we've not yet encountered a scenario where this (or any other) mutation has been necessary for an object that's not a first-order graph entity (that is, node or edge).

The one caveat I'd put on this statement is that all `RETURN` projections inherently hold identifiers, while `ORDER BY` projections do not. See the constructed identifier at `12`:
```
 @0   0..38  statement                  body=@1
 @1   0..38  > query                    clauses=[@2, @7]
 @2   0..10  > > MATCH                  pattern=@3
 @3   6..9   > > > pattern              paths=[@4]
 @4   6..9   > > > > pattern path       (@5)
 @5   6..9   > > > > > node pattern     (@6)
 @6   7..8   > > > > > > identifier     `a`
 @7  10..38  > > RETURN                 projections=[@8], ORDER BY=@13
 @8  17..23  > > > projection           expression=@9, alias=@12
 @9  17..22  > > > > apply              @10(@11)
@10  17..19  > > > > > function name    `ID`
@11  20..21  > > > > > identifier       `a`
@12  17..23  > > > > identifier         `ID(a)`
@13  23..38  > > > ORDER BY             items=[@14]
@14  32..38  > > > > sort item          expression=@15, ASCENDING
@15  32..37  > > > > > apply            @16(@17)
@16  32..34  > > > > > > function name  `ID`
@17  35..36  > > > > > > identifier     `a`
```

I think this discrepancy (provided it's not deliberate) would be better resolved in construction rather than mutation, however. In the same vein, we have tested a variant of constructors that build identifiers for unaliased graph entities upon construction, which is an approach that could be electively adopted as a parse flag much like `CYPHER_PARSE_ONLY_STATEMENTS`.

**Comment by cleishm** [comment]
By "alternative data structure", I meant one that holds annotations to the AST. Taking that approach, I could see us adding a mutable pointer to every AST node, allowing anyone to attach whatever data they like to any node of the tree.

**Comment by cleishm** [comment]
> I think this discrepancy (provided it's not deliberate)

It's quite deliberate. `RETURN` clauses create identifiers based on the name of the projection. These are used in the output field names, and also for the `UNION` clause. And when there is no `AS <identifier>` as part of the projection, the new identifier name is based on the stringified expression (hence `ID(a)` in your example). In contrast, `ORDER BY` is only ordering outputs and not creating projections.

**Comment by cleishm** [comment]
> build identifiers for unaliased graph entities upon construction, which is an approach that could be electively adopted as a parse flag much like `CYPHER_PARSE_ONLY_STATEMENTS`.

That's an interesting idea. However, the AST is meant to represent the syntax of the input text - and generating identifiers would change that semantic. I'm not saying that's a deal-breaker, it's just an interesting aspect to consider. I _think_ I prefer that to direct mutation, if for no other reason than it constrains the problem to this one issue (whereas adding one way to mutate the AST suggests a consistent API would have mutation methods for everything). I'd love to see a PR for this idea, if it's easy to send in.

We could perhaps add both this AND a single mutable pointer to every AST node, making it easy to annotate the tree for any processing purpose someone may need.

**Comment by jeffreylovitz** [comment]
Thanks for the clarification about identifiers in `RETURN` projections vs `ORDER BY` sort items!

Allowing for AST annotation via mutable pointers would be a valuable feature, I think, but (given my limited understanding of the code base) it seems like we'd have to write a lot of setter functions that mirror the current in situ logic. Unless there is some way to introduce mutations on the `cypher_astnode_t` interface rather than explicitly extending the interface of each type, but even that would be a significant deviation from the current API. (This was a bit of a ramble, but I'd love to hear your thoughts on how AST annotations could be implemented!)

In any case, I'd be happy to draft a PR that invokes the flag strategy discussed here! While it does deviate from the standard practice of the AST, that seems permissible to me as an opt-in flag, especially since I do believe that generated identifiers offer a lot of utility.

**Comment by cleishm** [comment]
@jeffreylovitz I just noticed another curiosity - the API does not currently support walking a mutable tree. All the `_get_` methods, e.g. `cypher_ast_query_get_clause(...)`, return a `const cypher_astnode_t *`.

The methods proposed in this PR take non-const pointers: `cypher_ast_node_pattern_set_identifier(cypher_astnode_t *node, cypher_astnode_t *identifier)`.

In using those methods, where do you obtain the non-const pointer from? The only methods that return non-const pointers to astnodes are the constructors - they cant be obtained as a result of parsing.

**Comment by jeffreylovitz** [comment]
We directly invoke the `cypher_ast_identifier()` constructor to build our non-const `identifier` argument, which I believe is fine. The only oddity there is the `range` argument, but I believe this does not have any consequence after AST construction!

You're right about the `node` arguments, which are in fact retrieved with the normal `_get_` methods and subsequently cast to remove `const`. This is definitely a violation.

I hope to have the flag PR ready early next week, which I believe will sidestep this issue!

**Comment by cleishm** [comment]
Right - so you're casting `const cypher_astnode_t *` back to a non-const `cypher_astnode_t *`. That's certainly not something I'd want to encourage as official API usage :)

**Comment by jeffreylovitz** [comment]
Haha, I definitely concede that!

**Comment by cleishm** [comment]
@jeffreylovitz I opened #22, which provides an annotation API you could use for this as well. Instead of directly updating the astnode, you'd attach new identifiers as annotations - which would then be available in your later processing.

The downside is that your code would need to look for the annotation and not just at the node's identifier, but it avoids all the non-const casting and need for a mutable tree.

**Comment by jeffreylovitz** [comment]
@cleishman The annotation API looks great, thanks very much for putting it together! We could definitely use this for our purposes.

Will there be an API for retrieving annotation contexts, or is the caller responsible for tracking those?

**Comment by cleishm** [comment]
My intention was to make the caller responsible for managing the lifecycle of the contexts, including keeping track of them.
